File: things/src/main/java/com/krisbiketeam/smarthomeraspbpi3/driver/MCP23017Pin.java
package com.krisbiketeam.smarthomeraspbpi3.driver;

@SuppressWarnings("WeakerAccess")
public class MCP23017Pin {

    public enum PinMode {
        DIGITAL_INPUT, DIGITAL_OUTPUT
    }
    public enum PinState {
        LOW, HIGH
    }

    public enum PinPullResistance {
        OFF, PULL_UP
    }

    public static final MCP23017Pin GPIO_A0 = new MCP23017Pin(1, "GPIO A0");
    public static final MCP23017Pin GPIO_A1 = new MCP23017Pin (2, "GPIO A1");
    public static final MCP23017Pin GPIO_A2 = new MCP23017Pin (4, "GPIO A2");
    public static final MCP23017Pin GPIO_A3 = new MCP23017Pin (8, "GPIO A3");
    public static final MCP23017Pin GPIO_A4 = new MCP23017Pin (16, "GPIO A4");
    public static final MCP23017Pin GPIO_A5 = new MCP23017Pin (32, "GPIO A5");
    public static final MCP23017Pin GPIO_A6 = new MCP23017Pin (64, "GPIO A6");
    public static final MCP23017Pin GPIO_A7 = new MCP23017Pin (128, "GPIO A7");
    public static final MCP23017Pin GPIO_B0 = new MCP23017Pin (1001, "GPIO B0");
    public static final MCP23017Pin GPIO_B1 = new MCP23017Pin (1002, "GPIO B1");
    public static final MCP23017Pin GPIO_B2 = new MCP23017Pin (1004, "GPIO B2");
    public static final MCP23017Pin GPIO_B3 = new MCP23017Pin (1008, "GPIO B3");
    public static final MCP23017Pin GPIO_B4 = new MCP23017Pin (1016, "GPIO B4");
    public static final MCP23017Pin GPIO_B5 = new MCP23017Pin (1032, "GPIO B5");
    public static final MCP23017Pin GPIO_B6 = new MCP23017Pin (1064, "GPIO B6");
    public static final MCP23017Pin GPIO_B7 = new MCP23017Pin (1128, "GPIO B7");

    public static final MCP23017Pin[] ALL_A_PINS = { MCP23017Pin.GPIO_A0, MCP23017Pin.GPIO_A1,
            MCP23017Pin.GPIO_A2, MCP23017Pin.GPIO_A3,
            MCP23017Pin.GPIO_A4, MCP23017Pin.GPIO_A5, MCP23017Pin.GPIO_A6, MCP23017Pin.GPIO_A7 };

    public static final MCP23017Pin[] ALL_B_PINS = { MCP23017Pin.GPIO_B0, MCP23017Pin.GPIO_B1,
            MCP23017Pin.GPIO_B2, MCP23017Pin.GPIO_B3,
            MCP23017Pin.GPIO_B4, MCP23017Pin.GPIO_B5, MCP23017Pin.GPIO_B6, MCP23017Pin.GPIO_B7 };

    public static final MCP23017Pin[] ALL = { MCP23017Pin.GPIO_A0, MCP23017Pin.GPIO_A1,
            MCP23017Pin.GPIO_A2, MCP23017Pin.GPIO_A3,
            MCP23017Pin.GPIO_A4, MCP23017Pin.GPIO_A5, MCP23017Pin.GPIO_A6, MCP23017Pin.GPIO_A7,
            MCP23017Pin.GPIO_B0, MCP23017Pin.GPIO_B1, MCP23017Pin.GPIO_B2, MCP23017Pin.GPIO_B3,
            MCP23017Pin.GPIO_B4, MCP23017Pin.GPIO_B5, MCP23017Pin.GPIO_B6, MCP23017Pin.GPIO_B7 };

    private final int address;
    private final String name;
    
    public MCP23017Pin(int address, String name) {
        this.address = address;
        this.name = name;
    }

    public String getName() {
        return name;
    }
    public int getAddress() {
        return address;
    }



}

File: things/src/main/java/com/krisbiketeam/smarthomeraspbpi3/driver/MCP23017Pin.kt
package com.krisbiketeam.smarthomeraspbpi3.driver

class MCP23017Pin {

    interface MCP23017PinStateChangeListener {
        fun onPinStateChanged(pin: Pin, state: PinState)
    }

    enum class PinMode {
        DIGITAL_INPUT, DIGITAL_OUTPUT
    }

    enum class PinState {
        LOW, HIGH
    }

    enum class PinPullResistance {
        OFF, PULL_UP
    }

    enum class Pin(var address: Int, var gpio_name: String) {
        GPIO_A0(1, "GPIO A0"),
        GPIO_A1(2, "GPIO A1"),
        GPIO_A2(4, "GPIO A2"),
        GPIO_A3(8, "GPIO A3"),
        GPIO_A4(16, "GPIO A4"),
        GPIO_A5(32, "GPIO A5"),
        GPIO_A6(64, "GPIO A6"),
        GPIO_A7(128, "GPIO A7"),
        GPIO_B0(1001, "GPIO B0"),
        GPIO_B1(1002, "GPIO B1"),
        GPIO_B2(1004, "GPIO B2"),
        GPIO_B3(1008, "GPIO B3"),
        GPIO_B4(1016, "GPIO B4"),
        GPIO_B5(1032, "GPIO B5"),
        GPIO_B6(1064, "GPIO B6"),
        GPIO_B7(1128, "GPIO B7")
    }

    companion object {

        val ALL_A_PINS = arrayOf(Pin.GPIO_A0, Pin.GPIO_A1, Pin.GPIO_A2, Pin.GPIO_A3, Pin.GPIO_A4,
                Pin.GPIO_A5, Pin.GPIO_A6, Pin.GPIO_A7)

        val ALL_B_PINS = arrayOf(Pin.GPIO_B0, Pin.GPIO_B1, Pin.GPIO_B2, Pin.GPIO_B3, Pin.GPIO_B4,
                Pin.GPIO_B5, Pin.GPIO_B6, Pin.GPIO_B7)

        val ALL = arrayOf(Pin.GPIO_A0, Pin.GPIO_A1, Pin.GPIO_A2, Pin.GPIO_A3, Pin.GPIO_A4,
                Pin.GPIO_A5, Pin.GPIO_A6, Pin.GPIO_A7, Pin.GPIO_B0, Pin.GPIO_B1, Pin.GPIO_B2,
                Pin.GPIO_B3, Pin.GPIO_B4, Pin.GPIO_B5, Pin.GPIO_B6, Pin.GPIO_B7)
    }
}